[Merged by Bors] - feat(MeasureTheory/Measure/TypeClass/NoAtoms): add `exists_accPt_of_noAtoms`

tdwag123 opened the pull request:
feat(MeasureTheory/Measure/TypeClass/NoAtoms): Added a theorem that states If a set has positive measure under an atomless measure, then it has an accumulation point.

Added a lemma in (Topology/DiscreteSubset): If a subset of a topological space has no accumulation points,
then it carries the discrete topology.

Theorem added: `exists_accPt_of_pos_hausdorffMeasure`
Lemma added: `discreteTopology_of_noAccPts `

**Harmonic's Aristotle gave the initial version of the proofs.** I did substantial work shortening the proof from 40 lines and refactoring it into another Lemma.

Special thanks to @plp127 and @CoolRmal for the useful feedback in the PR process.

Co-authored-by: @Aristotle-Harmonic



---
<!-- Your PR title will become the first line of the commit message.

In this box, the text above the `---` (if not empty) will be appended
to the commit message, and can be used to give additional context or
details. Please leave a blank newline before the `---`, otherwise GitHub
will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

When merging, all the commits will be squashed into a single commit
listing all co-authors.

Co-authors in the squash commit are gathered from two sources:

First, all authors of commits to this PR branch are included. Thus,
one way to add co-authors is to include at least one commit authored by
each co-author among the commits in the pull request. If necessary, you
may create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

Second, co-authors can also be listed in lines at the very bottom of
the commit message (that is, directly before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines
at the bottom of the commit message (before the `---`, and also before
any "Co-authored-by" lines) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

tdwag123 committed:
feat(add theorem exists_accPt_of_pos_hausdorffMeasure)

Added a theorem that states if a set has positive s-dimensional Hausdorff measure, then it has an accumulation point, along with necessary proofs.

tdwag123 committed:
Merge pull request #1 from tdwag123/tdwag123-patch-1

feat(MeasureTheory/Measure/Hausdorff)  ∀ 𝑠 > 0, ∀ 𝐸 ⊂ 𝑅^𝑛, 𝐻^𝑠 (𝐸)>0 ⇒ ∃ 𝑥 ∈ 𝑅^𝑛 , 𝑥 is an accumulation point of 𝑃(𝐸).

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tdwag123 committed:
Merge branch 'master' into master

tdwag123 committed:
Remove simp

I removed the simp using simp? to find the appropriate simp only statement

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tdwag123 committed:
Merge branch 'master' into master

CoolRmal commented on Mathlib/MeasureTheory/Measure/Hausdorff.lean:
```diff
@@ -1107,6 +1107,64 @@ theorem hausdorffMeasure_orthogonalProjection_le [RCLike 𝕜]
+      ENNReal.tsum_const, ENat.card_eq_top_of_infinite, ENat.toENNReal_top, iInf_pos,
+      nonpos_iff_eq_zero, mul_eq_zero, ENNReal.top_ne_zero, ENNReal.rpow_eq_zero_iff, true_and,
+      ENNReal.zero_ne_top, false_and, or_false, false_or]
+    exact hs
```
```suggestion
    apply iInf_le_of_le fun _ => {x}
    simpa using hs
```

CoolRmal commented on Mathlib/MeasureTheory/Measure/Hausdorff.lean:
```diff
@@ -1107,6 +1107,64 @@ theorem hausdorffMeasure_orthogonalProjection_le [RCLike 𝕜]
+        simp_all only
+      · intro a
+        subst a
+        simp_all only⟩
```
I think you should use refine here and the code after `by` should start in a new line. Also you should always use `simp_all` to end a goal instead of `simp_all only`.

CoolRmal commented on Mathlib/MeasureTheory/Measure/Hausdorff.lean:
```diff
@@ -1107,6 +1107,64 @@ theorem hausdorffMeasure_orthogonalProjection_le [RCLike 𝕜]
+/-- If a set has positive `s`-dimensional Hausdorff measure, then it has an accumulation point. -/
+theorem exists_accPt_of_pos_hausdorffMeasure {n : ℕ} {s : ℝ} {E : Set (EuclideanSpace ℝ (Fin n))}
+    [MeasurableSpace (EuclideanSpace ℝ (Fin n))] [BorelSpace (EuclideanSpace ℝ (Fin n))]
+    (hs : 0 < s) (hE : 0 < μH[s] E) : ∃ x, AccPt x (𝓟 E) := by
```
You want to be as general as you can when you contribute to Mathlib. Here instead of working with Euclidean spaces, you can instead consider a type `X` with instances `[MetricSpace X] [MeasurableSpace X] [BorelSpace X][LindelofSpace X]` and the rest of your code should still work.

plp127 replied:
I think it's more common to assume `TopologicalSpace.SeparableSpace` (you can even assume `TopologicalSpace.IsSeparable E`) or in these measure theory files I see `SecondCountableTopology` a lot, they're equivalent in the case of a metric space.

CoolRmal replied:
Yes I believe `TopologicalSpace.IsSeparable E` is a better assumption here.

tdwag123 committed:
Generalize exists_accPt_of_pos_hausdorffMeasure theorem

Refactor the theorem to use a more general type for the set and include separability condition. I tried to take into account the suggestions, but I may have missed some. The theorem needed some bigger reworking after the generalization

plp127 commented on Mathlib/MeasureTheory/Measure/Hausdorff.lean:
```diff
@@ -1107,6 +1107,49 @@ theorem hausdorffMeasure_orthogonalProjection_le [RCLike 𝕜]
+    have hsepE : SeparableSpace E := h_sep.separableSpace
+    simpa using (TopologicalSpace.separableSpace_iff_countable (α := E)).1 hsepE
+  have h_hausdorff_zero : ∀ x ∈ E, μH[s] {x} = 0 := fun x _ => by
+    haveI := noAtoms_hausdorff X hs
```
This isn't specific to the Hausdorff measure, right? This should work for any measure with no atoms, since this is the only place you use the fact that your measure is the Hausdorff measure.

tdwag123 replied:
Should I change the theorem statement for this? If so, where should I move the file, since it is no longer related to Hausdorff

tdwag123 committed:
Add theorem for accumulation points in atomless measures,

generalization and change of location

Added a theorem to show that if a set has positive measure under an atomless measure, then it has an accumulation point.

tdwag123 committed:
Revert

tdwag123 committed:
Refactor exists_accPt_of_noAtoms theorem 

+ found a simplification

tdwag123 committed:
Introduce lemma for discrete topology with no AccPts

Added lemma for discrete topology based on no accumulation points.

tdwag123 committed:
Merge pull request #2 from tdwag123/tdwag123-patch-3

Introduce lemma for discrete topology with no AccPts

tdwag123 committed:
Delete discreteTopology_of_noAccPts lemma

Removed lemma about discrete topology for sets with no accumulation points.

tdwag123 committed:
Import DiscreteSubset in NoAtoms module

CoolRmal reviewed (commented):
I think this is ready to be merged after some minor changes.

tdwag123 committed:
Remove redundant `TopologicalSpace` pt 1

Co-authored-by: Yongxi (Aaron) Lin <97214596+CoolRmal@users.noreply.github.com>

tdwag123 committed:
Remove redundant `TopologicalSpace` pt 2

Co-authored-by: Yongxi (Aaron) Lin <97214596+CoolRmal@users.noreply.github.com>

tdwag123 committed:
Refactor exists_accPt_of_noAtoms theorem + move

Moved out of the partial ordering section, factored some variables out the `exists_accPt_of_noAtoms ` theorem

tdwag123 committed:
Section wrappers around theorem

added section wrappers around the theorem to prevent issues with `μ : Measure α` versus the partial measure `α`

tdwag123 committed:
Refactor `exists_accPt_of_noAtoms` theorem

Refactor the `exists_accPt_of_noAtoms` theorem for clarity and structure.

-Removed the section for the theorem
-Changed the assumptions to make them weaker
-Changed where the open statements are

Still to be done:
See if I can refactor and make the assumptions even weaker.

tdwag123 committed:
Merge branch 'leanprover-community:master' into master

tdwag123 committed:
Weakens assumptions for `exists_accPt_of_noAtoms`

Changed the assumptions for `exists_accPt_of_noAtoms` from [EPseudometricSpace] to [TopologicalSpace X]

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

tdwag123 committed:
Merge branch 'master' into master

tdwag123 committed:
Merge branch 'master' into master

tdwag123 committed:
Refactored Mathlib/Topology/DiscreteSubset.lean

Used filters to shorten, as suggested byPatrickMassot

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

tdwag123 committed:
Merge branch 'master' into master

tdwag123 committed:
Merge branch 'master' into master
